### PR TITLE
schema: enum builder default value setter

### PR DESCRIPTION
**Pull request by darioielardi**
Enum field builder should have a Default() setter, as stated in the inline docs example:
```
field.Enum("state").
	Values(
		"on",
		"off",
	).
	Default("on")
```
but it's actually missing.

### CODE REVIEW

**Comment by a8m** [comment]
Hey @darioielardi, thanks for your contribution. 

Can you please add UT and integration tests for it? (codegen should validate that the given default is one of the values).

Thanks

**Comment by darioielardi** [comment]
Right I'll add tests for everything, should I open another pull request for json default value setter?

**Comment by a8m** [comment]
We intentionally disabled the default value option for JSON. Mainly, because the lack of type safety.

**Comment by darioielardi** [comment]
I removed the JSON part, and wrote UT for the enum default value. Writing integration tests I found out that the code generated for the default value ( the constant ) has two problems with enums:

1. type assertion fails because the field type is a custom generated one and the value passed to the Default() setter is a string.
2. Injecting $f.Type in the template results in an import of the current package, because the type is stored in TypeInfo as packageName.Type.

example const generated in `mymodule/ent/user/user.go`: 
```go
import (
  ...
  "mymodule/ent/user"
  ...
)

const (
  ..
  DefaultRole = descRole.Default.(user.Role)
  ...
)
```

First problem is solved with a `string` type assertion and then a conversion to the real generated enum type:

```go
{{ $default }} = {{ $f.Type }}({{ $desc }}.Default.(string))
```
but it still injects the type with the package name:

```go
DefaultRole = user.Role(descRole.Default.(string))
```

I can't figure out a good way to solve that. Any ideas?

**Comment by darioielardi** [comment]
Didn't realize the solution was under my eyes, everything is now solved and I'm just writing integration tests.

**Comment by darioielardi** [comment]
Everything should be done, you can now review.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+.idea/
```
Comment (a8m) [inline .gitignore]: Please remove this file. 

You can use [global gitignore](https://help.github.com/en/github/using-git/ignoring-files) for cases like these.

File: entc/gen/template/meta.tmpl [hunk 2]
```diff
@@ -58,8 +58,14 @@ var (
 		{{ end -}}
 		{{ if $f.Default }}
 			{{- $default := $f.DefaultName -}}
-			// {{ $default }} holds the default value on creation for the {{ $f.Name }} field.
-			{{ $default }} = {{ $desc }}.Default.({{ if $f.IsTime }}func() {{ end }}{{ $f.Type }})
+			{{ if $f.IsEnum }}
+			    {{ $enum := trimPackage $f.Type.String $.Package }}
+			    // {{ $default }} holds the default value on creation for the {{ $f.Name }} field.
+			    {{ $default }} = {{ $enum }}({{ $desc }}.Default.(string))
```
Comment (a8m) [inline entc/gen/template/meta.tmpl]: I'll prefer to use one of the constants that are defined in this file (yeah, enum is a pretty special here).

File: entc/load/schema.go [hunk 3]
```diff
@@ -37,6 +37,7 @@ type Field struct {
 	Tag           string          `json:"tag,omitempty"`
 	Size          *int64          `json:"size,omitempty"`
 	Enums         []string        `json:"enums,omitempty"`
+	EnumDefault   string          `json:"enum_default,omitempty"`
```
Comment (a8m) [inline entc/load/schema.go]: I don't think a new field is required here. Maybe we can just change the `Default` field to be type `interface{}`.

Let me know if you are fine with this, and I can help with the implementation part.

File: .gitignore (hunk 1 above)
Comment (darioielardi) [reply .gitignore]: yeah I'm sorry, that was unintentional

File: entc/load/schema.go (hunk 3 above)
Comment (darioielardi) [reply entc/load/schema.go]: As far as I understand right now the `load.Field.Default` is a bool and is used to know if that field has a default value set, it does not hold the actual value, right? That's why I came up with that `EnumDefault`, which is not the best solution, I agree, but I didn't want to mess up with the whole thing.

File: entc/gen/template/meta.tmpl (hunk 2 above)
Comment (darioielardi) [reply entc/gen/template/meta.tmpl]: I'm afraid I don't get what you mean here.

**Comment by a8m** [comment]
Hey, I didn't forget this diff.
I want to push it, however, there are some changes that need to be done first.

I'll get to it in the upcoming days.

**Comment by darioielardi** [comment]
No worries, feel free to answer my questions when you got the time.

**Comment by a8m** [comment]
Hey, I've added some changes to the `load` package (with default values) that can help you.
Let me know if you want to rebase and take it from there, or want me to pull and work on top of your changes. 

Thanks!

**Comment by darioielardi** [comment]
I'd probably have to wait some weeks to find the time to work on this, so it's probably better if you pull and work yourself on top of my changes, thank you.

**Comment by a8m** [comment]
I pulled this PR and rebased this on master (created a separated PR with your commit #271) . I'll merge this soon. 

Closing this one. 

Thanks for your contribution!

**Pull request closed without merging**